openerp/addons/base/ir/ir_qweb/qweb.py

```diff
@@ -0,0 +1,868 @@
+        """ returns a :class:`~openerp.models.Model` used to render a
+        ``t-field``.
+        By default, tries to get the model named
+        :samp:`ir.qweb.field.{field_type}`, falling back on ``ir.qweb.field``.
+        :param str field_type: type or widget of field to render
+        :param cursor cr
+        :param int uid
+        :param dict context
+        """
+        model = 'ir.qweb.field.' + field_type
+        env = self.env(cr=cr, user=uid, context=context)
+        return env[model] if model in env else env['ir.qweb.field']
+
+    def build_text(self, tagName, text):
+        return utils.unicodifier(text)
+
+    def build_attribute(self, tagName, name, value, qwebcontext):
+        return u' %s="%s"' % (name, openerp.tools.html_escape(utils.unicodifier(value)))
+
+
+    def format(value, formating, *args, **kwargs):
+        format = getattr(self, '_format_func_%s' % formating, None)
+        if not format:
+            raise "Unknown formating '%s'" % (formating,)
+        return format(value, *args, **kwargs)
+
+    def _format_func_monetary(self, value, display_currency=None, from_currency=None):
+        precision = int(round(math.log10(display_currency.rounding)))
+        fmt = "%.{0}f".format(-precision if precision < 0 else 0)
+        lang_code = self.env.context.get('lang') or 'en_US'
+        lang = self.env['res.lang']._lang_get(lang_code)
+        formatted_amount = lang.format(fmt, value, grouping=True, monetary=True)
+        pre = post = u''
+        if display_currency.position == 'before':
+            pre = u'{symbol}\N{NO-BREAK SPACE}'
+        else:
+            post = u'\N{NO-BREAK SPACE}{symbol}'
+        return u'{pre}{0}{post}'.format(
+            formatted_amount, pre=pre, post=post
+        ).format(symbol=display_currency.symbol,)
+
+
+    _void_elements = frozenset([
+        'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'keygen',
+        'link', 'menuitem', 'meta', 'param', 'source', 'track', 'wbr'])
+
+    _name_gen = itertools.count()
+
+    def compile(self, element):
+        xmlid = element.get('t-name', 'unknown')
+        mod = utils.base_module()
+
+        string = etree.tostring(element)
+
+        try:
+            ast_calls = self._compile_document(element)
+        except Exception, e:
+            raise QWebException("%s%s\nError when compiling AST" % (e, traceback.format_exc()), template=xmlid)
+
+        mod.body.extend(ast_calls)
+        ast.fix_missing_locations(mod)
+
+        def exeption(e, message):
+            if isinstance(e, QWebException):
+                return e
+            code = "Please install astor to display the compiled code"
+            stack = traceback.format_exc()
+            try:
+                import astor
+                code = astor.to_source(mod)
+            except ImportError:
+                stack += "\nFor more Error informations install astor lib"
+                pass
```

Comment:
useless `pass` statement, and optional imports should be done in the imports section e.g. 
```python
try:
    import astor
except ImportError:
    astor = None
```
with this section becoming:
```python 
if astor:
    code = astor.to_source(mod)
else:
    code = "Please install astor to display the compiled code"
    stack += "\nInstall `astor` for compiled source information."
```